src/data/matrix/kronecker.lean

```diff
@@ -0,0 +1,319 @@
+/-
+Copyright (c) 2021 Filippo A. E. Nuccio. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Filippo A. E. Nuccio, Eric Wieser
+-/
+
+import data.matrix.basic
+import linear_algebra.tensor_product
+import ring_theory.tensor_product
+
+/-!
+# Kronecker product of matrices
+
+This defines the [Kronecker product](https://en.wikipedia.org/wiki/Kronecker_product).
+
+## Main definitions
+
+* `matrix.kronecker_map`: A generalization of the Kronecker product: given a map `f : α   → β → γ`
+  and matrices `A` and `B` with coefficients in `α` and `β`, respectively, it is defined as the
+  matrix with coefficients in `γ` such that
+  `kronecker_map f A B (i₁, i₂) (j₁, j₂) = f (A i₁ j₁) (B i₁ j₂)`.
+* `matrix.kronecker_map_linear`: when `f` is bilinear, so is `kronecker_map f`.
+
+## Specializations
+
+* `matrix.kronecker`: An alias of `kronecker_map (*)`. Prefer using the notation.
+* `matrix.kronecker_linear`: `matrix.kronecker` is bilinear
+
+* `matrix.kronecker_tmul`: An alias of `kronecker_map (⊗ₜ)`. Prefer using the notation.
+* `matrix.kronecker_tmul_linear`: `matrix.tmul_kronecker` is bilinear
+
+## Notations
+
+These require `open_locale kronecker`:
+
+* `A ⊗ₖ B` for `kronecker_map (*) A B`
+* `A ⊗ₖₜ B` and `A ⊗ₖₜ[R] B` for `kronecker_map (⊗ₜ) A B`
+
+-/
+
+namespace matrix
+
+open_locale matrix
+
+variables {R α α' β β' γ γ' : Type*}
+variables {l m n p : Type*} [fintype l] [fintype m] [fintype n] [fintype p]
+variables {l' m' n' p' : Type*} [fintype l'] [fintype m'] [fintype n'] [fintype p']
+
+section kronecker_map
+
+/-- Produce a matrix with `f` applied to every pair of elements from `A` and `B`. -/
+@[simp] def kronecker_map (f : α → β → γ) (A : matrix l m α) (B : matrix n p β) :
+  matrix (l × n) (m × p) γ
+| i j := f (A i.1 j.1) (B i.2 j.2)
+
+lemma kronecker_map_transpose (f : α → β → γ)
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f Aᵀ Bᵀ = (kronecker_map f A B)ᵀ :=
+ext $ λ i j, rfl
+
+lemma kronecker_map_map_left (f : α' → β → γ) (g : α → α')
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f (A.map g) B = kronecker_map (λ a b, f (g a) b) A B :=
+ext $ λ i j, rfl
+
+lemma kronecker_map_map_right (f : α → β' → γ) (g : β → β')
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f A (B.map g) = kronecker_map (λ a b, f a (g b)) A B :=
+ext $ λ i j, rfl
+
+lemma kronecker_map_map (f : α → β → γ) (g : γ → γ')
+  (A : matrix l m α) (B : matrix n p β) :
+  (kronecker_map f A B).map g = kronecker_map (λ a b, g (f a b)) A B :=
+ext $ λ i j, rfl
+
+lemma kronecker_map_zero_left [has_zero α] [has_zero γ] (f : α → β → γ) (hf : ∀ b, f 0 b = 0)
```

Comment:
Should there be `@[simp]`-lemma variants of this lemma (and the ones below) in the case that `f` is a bundled map that preserves sufficient structure?

Reply:
I think _this_ lemma should just be marked simp.